modules/apps/portal/portal-company-log/src/main/java/com/liferay/portal/company/log/internal/servlet/CompanyLogServlet.java

```diff
@@ -0,0 +1,254 @@
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.PropsKeys;
+import com.liferay.portal.kernel.util.PropsUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.PrintWriter;
+
+import javax.servlet.Servlet;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Hai Yu
+ */
+@Component(
+	enabled = false, immediate = true,
+	property = {
+		"osgi.http.whiteboard.servlet.name=com.liferay.portal.company.log.internal.servlet.CompanyLogServlet",
+		"osgi.http.whiteboard.servlet.pattern=/company-log/*",
+		"servlet.init.httpMethods=GET"
+	},
+	service = Servlet.class
+)
+public class CompanyLogServlet extends HttpServlet {
+
+	@Override
+	protected void doGet(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException, ServletException {
+
+		try {
+			PermissionChecker permissionChecker = _getPermissionChecker(
+				httpServletRequest);
+
+			String path = HttpComponentsUtil.fixPath(
+				httpServletRequest.getPathInfo());
+
+			String[] pathArray = StringUtil.split(path, CharPool.SLASH);
+
+			if (pathArray.length == 0) {
+				_listCompaniesLogFiles(
+					permissionChecker, httpServletRequest, httpServletResponse);
+			}
+			else if (pathArray.length == 2) {
+				_downloadLogFile(
+					permissionChecker, httpServletRequest, httpServletResponse,
+					pathArray);
+			}
+		}
+		catch (FileNotFoundException fileNotFoundException) {
+			if (_log.isWarnEnabled()) {
+				_log.warn(fileNotFoundException);
+			}
+
+			_portal.sendError(
+				HttpServletResponse.SC_NOT_FOUND, fileNotFoundException,
+				httpServletRequest, httpServletResponse);
+		}
+		catch (Exception exception) {
+			if (_log.isWarnEnabled()) {
+				_log.warn(exception);
+			}
+
+			_portal.sendError(
+				exception, httpServletRequest, httpServletResponse);
+		}
+	}
+
+	private void _downloadLogFile(
+			PermissionChecker permissionChecker,
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse, String[] pathArray)
+		throws Exception {
+
+		long companyId = GetterUtil.getLongStrict(pathArray[0]);
+
+		if (_companyLocalService.fetchCompanyById(companyId) == null) {
+			throw new NoSuchCompanyException(
+				"No Company exists with the primary key " + companyId);
+		}
+
+		if (!permissionChecker.isCompanyAdmin(companyId)) {
+			throw new PrincipalException.MustBeCompanyAdmin(
+				permissionChecker.getUserId());
+		}
+
+		String fileName = pathArray[1];
+
+		String path = StringBundler.concat(
+			StringUtil.replace(
+				PropsUtil.get(PropsKeys.LIFERAY_HOME), '\\', '/'),
```

Comment:
@yuhai @dantewang Let's try not hard code the path here, as it is defined in xml now, we can not stop people changing it.

Tina.